### ISSUE TITLE
Proposal: Add a button on "Enter API Link" box allowing a user to get API link.

### ISSUE THREAD

**Issue by SoftwareGuy** [KanColleTool/KanColleTool#129]
At the current moment, one needs to extract their API link from the DMM website which is a pain in the butt. The "Enter API Link" needs a button that says `What? I don't have one...` or `Get API Link`. This will basically be the equivalent of doing "Get API Link" in the viewer but helps newcomers to the tool get started rather quickly.

This is an oversight because since we added the `Get API Link` function, the end-user cannot do this unless they actually have a token and they have fed it to the viewer's popup box on first boot. Which means they are likely to get frustrated and go use that Windows-only Metro-themed viewer instead.

**Comment by uppfinnarn** [comment]
"Get API Link" (which is a terrible name for it in the first place) was kinda hastily hacked in by @Yukariin a while ago, the API link prompt doesn't actually make that much sense anymore after that…

Perhaps a better way would be to add a start page that displays if there's no API link, that shows instructions for extracting it manually or getting it via the viewer?

**Comment by SoftwareGuy** [comment]
One idea I had, but not sure how to set it up is this:

***If the user cancels the "Get API Link" prompt on first boot***
On the "cancel" button, make a call back to the QTWeb element (I don't know what it's called, QTWeb something) to load a welcome page (can be located here, or just a hard-coded HTML markup page). Disable the "Reset" function (since we don't have an API Key, DMM will likely throw you an errorcat). 

That way people can set up KCTV (KanColle TV woot!) and the Tool, then take the plunge and get an API key.

**Comment by sauyon** [comment]
Maybe we could automate the process? It shouldn't be that difficult to send the login/scrape the api link with the region cookies.

(I'm back! Turns out uni is incredibly boring!)

**Comment by SunDwarf** [comment]
You'd have to actually use a web browser to view it, as the link is generated via terrible JS.

**Comment by Yukariin** [comment]
@SunDwarf actually now viewer can do cookie hack and get new apilink from iframe.
We just need remove that start-message with api-link, and update startup page and refresh.

**Comment by SunDwarf** [comment]
Goddamn magic has advanced so much this year.

**Comment by Rhextian** [comment]
can we have coloured hp bar in kancolle tool depends on the ships damage status? like yellow hp bar for shouha, orange hp bar for chuuha and red hp bar for taiha

**Comment by SunDwarf** [comment]
That would go on a separate issue.

**Comment by uppfinnarn** [comment]
I had started working on this on the rewrite branch of the viewer, actually. It's not difficult at all, once you have some proper structure to it.

(Then life happened and it's now kinda sitting there half-done…)

**Comment by Yukariin** [comment]
@uppfinnarn umm, there are Qt 5.6 release soon that include new version of QtWebEngine that should support latest (pepper) flash on all main platforms.
But WebEngine network part is different from webkit and seems doesn't use QNAM at all. So may be we should wait a bit and try to port it to QtWebEngine?

**Comment by SoftwareGuy** [comment]
I'm all for Pepper Flash because let's face it - NPAPI is old and most browsers want it to die in a fire. I'm sure we can adapt what we have now (our interception stuff) to the new WebEngine system or whatever Qt 5.6 brings to the table.

**Comment by sauyon** [comment]
All the javascript to get the api link appears to be unminified, I'm going to go through it and attempt to reimplement their auth scheme.

**Comment by uppfinnarn** [comment]
Yeah, we should be able to just run the upstream script and catch what it embeds in the page! Using a throwaway Web Page object (with no viewport backing and plugins disabled), we should be able to easily just grab the src attribute and roll with it.

As for WebEngine, I looked at it a while ago, and it seems like it'd be a little clunky to implement interception with it - it's very black box-y. It does its own thing, which does make sense since it's embedding Chromium's web stack, but it'd mean we need to use an HTTP proxy somehow.

One way that could work is an idea I've been playing around with: gutting out the interception from the viewer, then replacing the tool with an [Electron](http://electron.atom.io)-based thingie whose backend runs an HTTP proxy and possibly a minimal web server, that serves an API for querying game data. The frontend would be a Ember/React/something app that queries the backend, and the viewer would simply proxy requests to the tool, which would do the actual work.

**Comment by SunDwarf** [comment]
On 21/12/15 15:47, uppfinnarn wrote:
> Yeah, we should be able to just run the upstream script and catch what it embeds in the page! Using a throwaway Web Page object (with no viewport backing and plugins disabled), we should be able to easily just grab the src attribute and roll with it.
> 
> As for WebEngine, I looked at it a while ago, and it seems like it'd be a little clunky to implement interception with it - it's very black box-y. It does its own thing, which does make sense since it's embedding Chromium's web stack, but it'd mean we need to use an HTTP proxy somehow.
> 
> One way that could work is an idea I've been playing around with: gutting out the interception from the viewer, then replacing the tool with an Electron <http://electron.atom.io>-based thingie whose backend runs an HTTP proxy and possibly a minimal web server, that serves an API for querying game data. The frontend would be a Ember/React/something app that queries the backend, and the viewer would simply proxy requests to the tool, which would do the actual work.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/KanColleTool/KanColleTool/issues/129#issuecomment-166335701>.
> 

I think I just got a buzzword bingo.

**Comment by uppfinnarn** [comment]
Thank you, that is totally contributing to this conversation.

**Comment by SunDwarf** [comment]
Glad I could be of service.
